### PR TITLE
RSDK-5317 - Provide default environent variables and a data directory for modules

### PR DESCRIPTION
**Pull request by zaporter-work**
Ticket: https://viam.atlassian.net/browse/RSDK-5317

This PR accomplishes two main tasks:
1) Provide the default module environment variables (excluding VIAM_MODULE_ROOT which will be done in its own special way). This is done via the addition of a `getFullEnvironment(..)` fn on `modmanager.Module` which is called when starting the module process
2) Create directories for modules to store permanent data. This is useful for modules that download data to the robot (like all python modules which can now store their venv's in $VIAM_MODULE_DATA so that it is persisted across module versions (and is cleaned up when the module is removed))
For all cloud robots, each module is given a data directory in the format:
$VIAM_HOME/module-data/\<cloud-robot-id\>/\<module-name\> (ex: /home/steve/.viam/module-data/aecf-2343-1e32/wifi-sensor)
For all local modules, the \<cloud-robot-id\> is replaced with "local" (ex: /home/steve/.viam/module-data/local/wifi-sensor)

This diverges slightly from the scope (which did not have the extra layer for the robot id) because I realized that people would be surprised/frustrated when their data directory was purged every time they switched robot configs and they had to wait for their modules to redownload data. This follows the pattern established with the `viam_cached_config_<cloud-robot-id>.json`

Questions for reviewers:
- ~Do you have any suggestions for injecting a `viamHomeDir` into the complex & simple module tests? Otherwise they use `$HOME/.viam` which is non-ideal because it means that anyone developing on the RDK (and who runs the tests on their host machine) will have their `$VIAM_HOME/module-data/local/...` directory cleared every time they run the tests. This isn't that bad... however it is a bit ugly.~ @dgottlieb Had the brilliant idea to just pass a new `$HOME` to the viam-server in the complex module tests. This fixed it. While doing that, I decided to also add some full e2e tests that the data directory is created and that the gizmo is able to write data to it.

### CODE REVIEW

File: config/module.go
Comment (zaporter-work) [inline config/module.go:66]: flyby

File: module/modmanager/manager_test.go
Comment (zaporter-work) [inline module/modmanager/manager_test.go:47]: Flyby. I missed this when I did https://github.com/viamrobotics/rdk/pull/2902

Might be worth it to double-check that this test passes on macOS. I think it will but I haven't tested.

File: module/modmanager/manager_test.go
Comment (zaporter-work) [inline module/modmanager/manager_test.go:238]: This single "test" called `TestModManagerFunctions` is pretty gross.

If people are interested, I'd be happy to break it out into many `t.Run(...)`s.

However, I'd rather to that in a different PR.

File: robot/impl/resource_manager.go
Comment (zaporter-work) [inline robot/impl/resource_manager.go:967]: match spacing of the rest of the function

**Comment by dgottlieb** [comment]
Still looking but figured I'd move the ball on this question:

> Do you have any suggestions for injecting a viamHomeDir into the complex & simple module tests? Otherwise they use $HOME/.viam which is non-ideal because it means that anyone developing on the RDK (and who runs the tests on their host machine) will have their $VIAM_HOME/module-data/local/... directory cleared every time they run the tests. This isn't that bad... however it is a bit ugly.

IIUC, the consequence is that running tests touches the "real" viam home directory. But it won't actually delete "real" data, just the directory that test uses? I think that's fine (but definitely a problem is running the test can wipe out my real data).

I suppose the reservation about injecting a `viamHoneDir` is we lose testing coverage of the default value? I assume you could try having the tests that matter set the `$HOME` env variable (and disallow them from running in parallel).

File: module/modmanager/manager.go
Comment (benjirewis) [inline module/modmanager/manager.go:158]: [q] Is this possible?

File: module/modmanager/manager.go
Comment (benjirewis) [inline module/modmanager/manager.go:188]:
```suggestion
			return errors.WithMessage(err, "error creating data directory for module "+mod.cfg.Name)
```

File: module/modmanager/manager.go
Comment (benjirewis) [inline module/modmanager/manager.go:914]: Aside but with all these env variables, this makes me wish we did module log levels as environment variables as @Otterverse originally suggested. At the least, would unify logic here instead of having all of this _and_ a command line argument for log level.

File: module/modmanager/manager.go
Comment (benjirewis) [inline module/modmanager/manager.go:539]: This does not need a context parameter.

File: module/modmanager/manager.go
Comment (benjirewis) [inline module/modmanager/manager.go:936]: What's the expected behavior here? That a test with no `ViamHomeDir` modmanager option will cause an error? Where?

**Comment by zaporter-work** [comment]
>  the consequence is that running tests touches the "real" viam home directory. But it won't actually delete "real" data, just the directory that test uses

Well.. it will delete real data that is in your $HOME/.viam/module-data/local directory, so that is kind of real data

The hard part is the that the "complex module" tests run the real viam-server so there is no great spot to inject a different directory.

I could add a hidden top level arg to viam-server but that feels a bit ugly.

File: module/modmanager/manager_test.go
Comment (benjirewis) [reply module/modmanager/manager_test.go:47]: Just checked; it passes on linux but hangs on MacOS and then fails for me, I'll follow up offline.

File: module/modmanager/manager_test.go
Comment (benjirewis) [reply module/modmanager/manager_test.go:238]: Up to you but I'd say defer to another PR. This is one of our only tests that has this `t.Log`-separating-sections design; I don't think it's necessarily worse that a bunch of `t.Run`s.

File: module/modmanager/manager_test.go
Comment (benjirewis) [inline module/modmanager/manager_test.go:241]: [nit] can we assert something about the error message.

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:101]: Can we document here the expected values this (and `modulateDataParentDirectory`) can take on a post-initialized `Manager` object? When I read variables like these in code I'm always wondering what is `openFile(str)` going to find?
* Can the string contain env variables (e.g: `$HOME`) or other things that the program will evaluate/substitute before passing to the OS?
* Can the string be a relative path? Is it relative to the process PWD (os evaluated)? Is it relative to some other directory-representing string (program evaluated)?
* Is the string guaranteed to be an absolute path?

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:102]: Probably want to align on using `Dir` or `Directory` for these adjacent variable names.

I'd probably vote for `Dir` here to have avoided the reflowing of every member in this structure, but the damage was already done -- I already scanned the diff line by line to convince myself these were the only two changes :)

File: module/modmanager/options/options.go
Comment (benjirewis) [inline module/modmanager/options/options.go:14]:
```suggestion
	// Root of Viam server configuration. This is generally either /opt/viam or ~/.viam
```
Just for my own understanding, this is expected to literally be `/opt/viam` or `~/.viam`, not a directory within those directories, correct?

**Review by benjirewis — CHANGES_REQUESTED** [review]
Looking good! Some qs and nits.

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:934]: `creating a file`

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:158]: If the conf name contains characters like ".." (which is technically not allowed in the regex but might in at some point in the future and people won't remember to add this check).

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:936]: The expected behavior is that if no `ViamHomeDir` is passed, we should return an empty module data parent directory which will not create any data directories. There are a multitude of modmanager tests that have nothing to do with these data directories.

I could create temp dirs for every one of them, but that feels unnecessary imo.

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:550]: It'd be good to have an info log saying we're removing (the value of) `moduleDataParentDirectory`.

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:568]: Similarly, logging this will be useful.

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:166]: Can we please log the value of `moduleDataDir`? Maybe it's not necessary if it's getting logged as part of starting a module process.

I just want to make sure that if module startup depends on finding something in its data dir, we don't have to guess what the program thinks is the data dir.

**Comment by benjirewis** [comment]
For the complex module/simple module injection question, if you haven't solved it or taken @dgottlieb's suggestion, you could also try modifying the `Environment` field of passed in managed process config [here](https://github.com/viamrobotics/rdk/blob/main/examples/customresources/demos/complexmodule/moduletest/module_test.go#L49C47-L54) to have a custom `VIAM_HOME` value.

File: module/modmanager/manager.go
Comment (benjirewis) [reply module/modmanager/manager.go:936]: Ah sounds good; thanks.

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:932]: Any commentary on what `Parent` means here? The return value on 942 includes `moduleDataFolderName` as a non-terminating directory. I would expect the phrase "module data parent directory" in this case to be `options.ViamHomeDir` as that's the parent of "module data folder name".

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:155]: Documenting the intent here will be useful, such as:
* Under what circumstances is this not empty
* The check on 158 seems like a tautology. `moduleDataDir` is constructed as a join of `moduleDataParentDirectory`. Explaining what the expected environmental inputs are such that `Clean` can result in not finding a prefix would help future maintainers.

File: module/modmanager/options/options.go
Comment (zaporter-work) [reply module/modmanager/options/options.go:14]: Correct -- clarified in comment

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:102]: Oops -- thanks for catching the inconsistency.

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:550]: Great suggestion.

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:166]: Adding debug log when creating the dir.

File: module/modmanager/manager.go
Comment (Otterverse) [reply module/modmanager/manager.go:914]: *doesn't say it* :-)

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:166]: Added.

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:571]: Added

File: module/modmanager/manager_test.go
Comment (zaporter-work) [reply module/modmanager/manager_test.go:241]: Resolving.

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go]: Good catch. Removed.

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go]: Renamed to `Dir`

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go]: I really should get my editor setup to warn me about grammar mistakes... :pray:

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go]: Added.

File: examples/customresources/apis/proto/api/component/gizmo/v1/gizmo.proto
Comment (benjirewis) [inline examples/customresources/apis/proto/api/component/gizmo/v1/gizmo.proto:29]: I'm hesitant about expanding the API for gizmo. I suppose it's somewhat interesting to show that modular resources can write to files, but I wonder if we could use the `testmodule`'s `helper` resource and some new custom command to test that writes to files happen in the expected directory. Do you have any thoughts @Otterverse ?

**Review by benjirewis — COMMENTED** [review]
MacOS still hanging + failing `TestModManagerFunctions`, I'll take a look locally and see if I can determine where.

File: examples/customresources/apis/proto/api/component/gizmo/v1/gizmo.proto
Comment (zaporter-work) [reply examples/customresources/apis/proto/api/component/gizmo/v1/gizmo.proto:29]: I debated this as well. I liked Gizmo because it seemed like a better full e2e test. However, I was on the fence and `testmodule` would mean changing fewer LOC which is always a good thing.

Comment (benjirewis) [reply examples/customresources/apis/proto/api/component/gizmo/v1/gizmo.proto:29]: Yeah I'd vote for using `testmodule` instead of expanding the gizmo API if that doesn't sound like too much work.

Comment (zaporter-work) [reply examples/customresources/apis/proto/api/component/gizmo/v1/gizmo.proto:29]: You got it. Switching now.

Comment (Otterverse) [reply examples/customresources/apis/proto/api/component/gizmo/v1/gizmo.proto:29]: Agree with benji.

File: module/modmanager/manager.go
Comment (benjirewis) [inline module/modmanager/manager.go:943]:
```suggestion
// This directory should contain exactly one directory for each module present on the modmanager
```

File: module/modmanager/manager.go
Comment (benjirewis) [reply module/modmanager/manager.go:950]: Also confused about what `Parent` means here.

**Review by benjirewis — COMMENTED** [review]
Just nits! `TestModManagerFunctions` passes on mac now, thanks 🧑‍🔧

File: module/modmanager/manager.go
Comment (dgottlieb) [reply module/modmanager/manager.go:194]: My preference is that logs are sentences. With capitalization and punctuation.

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:104]: IIUC, this should only be empty as a convenience for testing? If that's true, I think it's preferable to call that out:

```
- it is empty if the modmanageroptions.Options.viamHomeDir was empty
+ `viamHomeDir` may only be the empty string in testing
```

File: module/modmanager/manager.go
Comment (dgottlieb) [reply module/modmanager/manager.go:950]: I think I get the relationship now. I believe the directory layout of viam-server data is the following:
```
- viamHomeDir/ (e.g: /home/robot/.viam)
-- <stuff>
-- moduleDataParentDir/ (e.g: /home/robot/.viam/module-data -- stored on the modmanager)
--- module1Data/ (`module.dataDir` variable, passed to the module on startup)
---- <files>
--- module2Data/
---- <files>
--- module3Data/
---- <files>
```

So the "parent dir" is really the "root" directory where module data directories go. Or the parent of the yet to be evaluated module data dirs.

Except I spoke a white lie. Presumably because multiple viam-servers can run on the same machine (more likely with local developer workflows I presume), we actually have a "robot id" directory. So the path to a module is:
```
/home/robot/.viam/module-data/robot.id/module1Data/
```

I wonder if we should instead be swapping the robot id and module data:
```
/home/robot/.viam/robot.id/module-data/module1Data/
```

But definitely don't want to muck with any precedence there.

File: module/modmanager/manager.go
Comment (dgottlieb) [inline module/modmanager/manager.go:546]:
```suggestion
// CleanModuleDataDirectory removes unexpected folders and files from the robot's module data directory.
// Modules removed from the robot config (even temporarily) will get pruned here.
```

File: module/modmanager/manager.go
Comment (benjirewis) [reply module/modmanager/manager.go:194]: Sounds good; ignore my nits (I'll remove comments).

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:950]: >I wonder if we should instead be swapping the robot id and module data:
```
/home/robot/.viam/robot.id/module-data/module1Data/
```
But definitely don't want to muck with any precedence there.

That seems like a good idea at face value... but I haven't had time to think it though and that is decently far from both what was scoped and what the current pattern is in the `.viam` directory... However, open to it.


> So the "parent dir" is really the "root" directory where module data directories go. Or the parent of the yet to be evaluated module data dirs.

That is exactly correct. That is the relationship. The `moduleDataParentDirectory` is the parent to all of the module's `data_dir`s. 

Do you have a suggestion as to how I should name this to be less confusing? (because clearly it is confusing)

Comment (benjirewis) [reply module/modmanager/manager.go:950]: Gotcha; both your comments are helpful, thanks. I was confused because the word "parent" is used to describe the module manager's relationship to modules (modules are aware of a "parent address" that is the sock file the viam server/module manager is sending gRPC requests from). `moduleDataParentDirectory` seems fitting now that you two have described the file layout. I have no strong opinion on inverting robot ID and module data for the directory structure.

File: module/modmanager/manager.go
Comment (zaporter-work) [reply module/modmanager/manager.go:105]: Because the home dir is required for env var insertion. (and I would prefer to store the moduleDataParentDir instead of storing the robot id and re-evaluating the parentDir every time it is needed... that is a personal preference though and I am open to your stylistic opinions.)

File: examples/customresources/apis/proto/api/component/gizmo/v1/gizmo.proto
Comment (zaporter-work) [reply examples/customresources/apis/proto/api/component/gizmo/v1/gizmo.proto]: Resolving.

File: module/modmanager/manager_test.go
Comment (zaporter-work) [reply module/modmanager/manager_test.go]: Moved this to https://github.com/viamrobotics/rdk/pull/3202

**Review by benjirewis — APPROVED** [review]
LGTM assuming tests pass! Thanks for all the back and forth 🧑‍🔧 .

**Review by dgottlieb — APPROVED** [review]

**Pull request merged**
